src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,285 @@
+  this->criticUpdater.Initialize(criticNetwork.Parameters().n_rows,
+                                 criticNetwork.Parameters().n_cols);
+  #else
+  this->criticUpdatePolicy = new typename UpdaterType::template
+  Policy<arma::mat, arma::mat>(this->criticUpdater,
+                               criticNetwork.Parameters().n_rows,
+                               criticNetwork.Parameters().n_cols);
+  #endif
+
+  #if ENS_VERSION_MAJOR == 1
+  this->actorUpdater.Initialize(actorNetwork.Parameters().n_rows,
+                                actorNetwork.Parameters().n_cols);
+  #else
+  this->actorUpdatePolicy = new typename UpdaterType::template
+  Policy<arma::mat, arma::mat>(this->actorUpdater,
+                               actorNetwork.Parameters().n_rows,
+                               actorNetwork.Parameters().n_cols);
+  #endif
+
+  // Copy over the actor parameters to its older self.
+  oldActorNetwork.Parameters() = actorNetwork.Parameters();
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType
+>
+PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType
+>::~PPO()
+{
+  #if ENS_VERSION_MAJOR >= 2
+  delete actorUpdatePolicy;
+  delete criticUpdatePolicy;
+  #endif
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType
+>::SelectAction()
+{
+  // Get action logits for each action at current state.
+  arma::colvec actionLogit;
+  actorNetwork.Predict(state.Encode(), actionLogit);
+
+  arma::colvec prob;
+  ann::Softmax<> softmax;
+  softmax.Forward(actionLogit, prob);
+
+  // Calculating cumulative probablity.
+  for (size_t i = 1; i<prob.n_rows; i++)
+      prob[i] += prob[i-1];
+
+  // Sampling action from cumulative probablity.
+  double randValue = arma::randu<double>();
+  for (size_t i = 0; i<prob.n_rows; i++)
+      if (randValue <= prob[i])
+    {
+      action.action = static_cast<decltype(action.action)>(i);
+      return;
+    }  
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType
+>::Update()
+{
+  arma::mat sampledStates = state.Encode();
+  std::vector<ActionType> sampledActions = {action};
+  arma::rowvec sampledRewards = {reward};
+  arma::irowvec isTerminal = {done};
+  arma::mat criticGradients, actorGradients;
+
+  arma::rowvec actionValues, advantages;
+  criticNetwork.Forward(sampledStates, actionValues);
+
+  arma::rowvec discountedRewards(sampledRewards.n_cols);
+  arma::rowvec nextActionValue;
+  criticNetwork.Predict(nextState.Encode(), nextActionValue);
+  double R = nextActionValue(0);
+  for (size_t i = sampledRewards.n_cols; i > 0; --i)
+  {
+    if (isTerminal[i-1])
```

Comment:
```suggestion
    if (isTerminal[i - 1])
```

Minor style adjustment, using a space between operations.